gateway/src/endpoints/batch_inference.rs

```diff
@@ -0,0 +1,886 @@
+    pub function_name: &'a str,
+    pub variant_name: &'a str,
+    pub episode_ids: &'a Vec<Uuid>,
+    pub tags: Option<Vec<Option<HashMap<String, String>>>>,
+    // pub tool_configs: &'a Vec<Option<ToolCallConfig>>,
+}
+
+#[derive(Debug, Serialize)]
+struct BatchModelInferenceInsert<'a> {
+    pub inference_id: String,
+    pub batch_id: &'a str,
+    pub function_name: &'a str,
+    pub variant_name: &'a str,
+    pub episode_id: String,
+    pub input: String,
+    pub input_messages: String,
+    pub system: Option<&'a str>,
+    pub tool_params: Option<ToolCallConfigDatabaseInsert>,
+    pub inference_params: &'a InferenceParams,
+    pub output_schema: Option<String>,
+    pub model_name: &'a str,
+    pub model_provider_name: &'a str,
+    pub tags: HashMap<String, String>,
+}
+
+#[derive(Debug, Serialize)]
+struct BatchRequestInsert<'a> {
+    batch_id: &'a str,
+    id: String,
+    batch_params: String,
+    model_name: &'a str,
+    model_provider_name: &'a str,
+    status: BatchStatus,
+    errors: HashMap<String, String>,
+}
+
+impl<'a> BatchRequestInsert<'a> {
+    fn new(
+        batch_id: &'a str,
+        batch_params: Value,
+        model_name: &'a str,
+        model_provider_name: &'a str,
+        status: BatchStatus,
+        errors: Option<HashMap<String, String>>,
+    ) -> Self {
+        let id = Uuid::now_v7().to_string();
+        let errors = errors.unwrap_or_default();
+        Self {
+            batch_id,
+            id,
+            batch_params: serde_json::to_string(&batch_params)
+                .map_err(|e| {
+                    Error::new(ErrorDetails::Serialization {
+                        message: e.to_string(),
+                    })
+                })
+                .unwrap_or_default(),
+            model_name,
+            model_provider_name,
+            status,
+            errors,
+        }
+    }
+}
+
+// Returns the batch ID and the inference IDs that were written to ClickHouse
+async fn write_inference<'a>(
+    clickhouse_connection_info: &ClickHouseConnectionInfo,
+    inputs: Vec<Input>,
+    result: BatchModelInferenceWithMetadata<'a>,
+    metadata: BatchInferenceDatabaseInsertMetadata<'a>,
+    inference_config: BatchInferenceConfig<'a>,
+) -> Result<(Uuid, Vec<Uuid>), Error> {
+    let mut rows = vec![];
+    let batch_id = result.batch_id.to_string();
+
+    for (
+        i,
+        inference_id,
+        input,
+        input_messages,
+        system,
+        tool_config,
+        inference_params,
+        output_schema,
+        tags,
+    ) in izip!(
+        0..,
+        result.inference_ids.iter(),
+        inputs,
+        result.input_messages.iter(),
+        result.systems.iter(),
+        inference_config.tool_configs,
+        result.inference_params.iter(),
+        result.output_schemas.iter(),
+        metadata
+            .tags
+            .unwrap_or_default()
+            .into_iter()
+            .chain(repeat(None)),
```

Comment:
Can you clean this up with a struct?

Reply:
done, still not that neat but at least we collect then serialize